pkg/observer/priority.go

```diff
@@ -0,0 +1,431 @@
+package observer
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"time"
+
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1 "github.com/atlassian-labs/cyclops/pkg/apis/atlassian/v1"
+	"github.com/atlassian-labs/cyclops/pkg/generation"
+)
+
+// recordPriorityMetrics records priority-related metrics
+func (c *controller) recordPriorityMetrics(priority int32, action string) {
+	c.PriorityActions.WithLabelValues(fmt.Sprintf("%d", priority), action).Inc()
+}
+
+// recordPriorityHealth records the health status of a priority level
+func (c *controller) recordPriorityHealth(priority int32, healthy bool) {
+	value := 0.0
+	if healthy {
+		value = 1.0
+	}
+	c.PriorityLevelHealth.WithLabelValues(fmt.Sprintf("%d", priority)).Set(value)
+}
+
+// recordPriorityActivationDuration records the time taken to activate a priority level
+func (c *controller) recordPriorityActivationDuration(priority int32, duration time.Duration) {
+	c.PriorityActivationDuration.WithLabelValues(fmt.Sprintf("%d", priority)).Observe(duration.Seconds())
+}
+
+// groupByPriority groups node groups by their priority
+func (c *controller) groupByPriority(changedNodeGroups []*ListedNodeGroups) map[int32][]*ListedNodeGroups {
+	priorityGroups := make(map[int32][]*ListedNodeGroups)
+
+	for _, nodeGroup := range changedNodeGroups {
+		priority := c.getPriority(nodeGroup.NodeGroup)
+		priorityGroups[priority] = append(priorityGroups[priority], nodeGroup)
+	}
+
+	return priorityGroups
+}
+
+// validatePriority validates that priority values are reasonable
+func (c *controller) validatePriority(priority int32) bool {
```

Comment:
I'd argue we don't need to worry about a limit tbh, I think it was be any integer, just order them and apply